### DISCUSSION TITLE
Update nuget please

### DISCUSSION THREAD

**Discussion by bjornbouetsmith** [DSharpPlus/DSharpPlus#807] [Q&A] [answered]
Hi,

Quoting from your from page:
"The latest stable release is always available on NuGet. This branch is less up-to-date than the nightly versions, but is guaranteed to not receive any breaking API changes without a major version bump."

The latest version in nuget is 3.2.3 - and was published 2017.

Surely some of the newer versions are considered stable?

**Comment by Emzi0767** [comment]
The term "stable" encompasses more than "it runs without crashes". Stability in this context also includes, among other things, API surface.

There are newer builds available through our nightly feed as well as the release candidates on NuGet. However these are constantly evolving and, as such, do not fit the definition of stable.

**Reply by bjornbouetsmith** [reply]
Sure I know what stable means - but it has been more than 3 years since the last "stable" release.

So perhaps its time to "release" something instead of keeping on polishing a product that is already being used in the wild many places.

I know it sucks to "lock" your API in place, but in more than 3 years, surely you should have decided on the final API?

**Reply by Emzi0767** [reply]
The problem isn't what we decide or not decide; you have to bear in mind that Discord themselves also changes things in a manner which often require consideration.

Moreover, we, the D#+ developers, are volunteers. We're not paid for maintaining the library, and we all have lives outside of maintaining the library. And time is a finite resource, which for some of us is rather scarce.

If you want to accelerate a release, you always have the option of browsing the issue tracker or asking in our discord server to see what work items are there still on the to-do list, and contributing your time and effort.

**Reply by bjornbouetsmith** [reply]
>> The problem isn't what we decide or not decide; you have to bear in mind that Discord themselves also changes things in a manner which often require consideration.

You decide when you want to stop implementing new features.

Discord decides when they release new versions of their API, but their API is backwards compatible I assume, since your code from 2017 still works.

The problem is that it seems like from the outside that you never stopped implementing new features, keeping up to date with discords API - which means if discord never stops deploying new API's you will never release.

So you should have designed your V4 API - and then made a CUT and said - "on this day - this version of the discord API is what we will use".

That would have enabled you to finalize/stabilize and finally release a V4 of your API - probably years ago.

And even if discord made 1k changes to their new API's it would not affect your release, since you were still using the old api.

And then when you had stabilized/released that - then you could have looked at "what is new" in discord api and enchanged, or even broken your API apart again.

You are also still updating packages in the version you are trying to stabilize - which means whenever their API changes you have to "fix" it - instead of having locked your version "on this day".
You are also still updating what discord API you are programming against - I know the latest is always the coolest, but you are shooting yourself in the feet when you change API in the middle of stabilizing and trying to release.

I know we all have finite time, but you could have saved yourself a lot of time and worry if you had locked dependencies a couple of years ago.

**Reply by Emzi0767** [reply]
The end result would instead be people complaining about missing features. There's nothing preventing you from using nightly builds or RCs.

**Reply by bjornbouetsmith** [reply]
>> The end result would instead be people complaining about missing features. 

And then you could have said - it comes in next version - come help us make the next version.

>> There's nothing preventing you from using nightly builds or RCs.

Nobody with any ounce of sanity will use nighly builds/RC's for their developments, since you have 0 guarantee that when you upgrade to "final" version that it has not all changed. Just because the maintainer decided to change something so it matches the new version of the API they are depending on.

You use a stable version to build on - make your own "release" - and push new features to next version.

**Reply by Emzi0767** [reply]
> Nobody with any ounce of sanity will use nighly builds/RC's for their developments, since you have 0 guarantee that when you upgrade to "final" version that it has not all changed.

Doesn't seem to bother most of the members of our community who run multi-dozen shard bots.

**Comment by sh30801** [comment, marked as answer]
Version 4.0 was released today 🎉
